toolkit/tools/imagegen/diskutils/verity.go

```diff
@@ -0,0 +1,316 @@
+// This map will have the correct name of the final verity disk, and can be
+// switched to read-only when the final image is ready for measurement.
+// - partDevPath is the path of the root partition device (likely a loopback device)
+// - partition is the disk configuration
+// - readOnlyConfig is the root read-only settings
+func PrepReadOnlyDevice(partDevPath string, partition configuration.Partition, readOnlyConfig configuration.ReadOnlyVerityRoot) (readOnlyDevice VerityDevice, err error) {
+	const (
+		linearTable = `0 %d linear %s 0`
+	)
+
+	if !readOnlyConfig.Enable {
+		err = fmt.Errorf("Verity is not enabled, can't update partition '%s'", partition.ID)
+		return
+	}
+
+	// Save the required information from the config into the device struct
+	finalDeviceName := fmt.Sprintf("%s%s", mappingVerityPrefix, readOnlyConfig.Name)
+	readOnlyDevice.BackingDevice = partDevPath
+	readOnlyDevice.MappedName = finalDeviceName
+	readOnlyDevice.MappedDevice = filepath.Join("/dev/mapper/", finalDeviceName)
+	readOnlyDevice.ErrorBehavior = readOnlyConfig.VerityErrorBehavior.String()
+	readOnlyDevice.ValidateOnBoot = readOnlyConfig.ValidateOnBoot
+	if readOnlyConfig.ErrorCorrectionEnable {
+		readOnlyDevice.FecRoots = readOnlyConfig.ErrorCorrectionEncodingRoots
+	} else {
+		readOnlyDevice.FecRoots = 0
+	}
+	readOnlyDevice.TmpfsOverlays = readOnlyConfig.TmpfsOverlays
+	if readOnlyConfig.TmpfsOverlayDebugEnabled {
+		readOnlyDevice.TmpfsOverlaysDebugMount = debugMountPoint
+	}
+	readOnlyDevice.UseRootHashSignature = readOnlyConfig.RootHashSignatureEnable
+
+	// linear mappings need to know the size of the disk in blocks ahead of time
+	deviceSizeStr, stderr, err := shell.Execute("blockdev", "--getsz", readOnlyDevice.BackingDevice)
+	if err != nil {
+		err = fmt.Errorf("unable to get loopback device size %s. Error: %s", partDevPath, stderr)
+		return
+	}
+	deviceSizeInt, err := strconv.ParseUint(strings.TrimSpace(deviceSizeStr), 10, 64)
+	if err != nil {
+		err = fmt.Errorf("Unable to convert disk size '%s' to integer: %w", deviceSizeStr, err)
+		return
+	}
+
+	populatedTable := fmt.Sprintf(linearTable, deviceSizeInt, readOnlyDevice.BackingDevice)
+	dmsetupArgs := []string{
+		"create",
+		readOnlyDevice.MappedName,
+		"--table",
+		populatedTable,
+	}
+	_, stderr, err = shell.Execute("dmsetup", dmsetupArgs...)
+	if err != nil {
+		err = fmt.Errorf("Unable to create a device mapper device '%s': %w", stderr, err)
+		return
+	}
+
+	logger.Log.Debugf("Remapped partition %s for read-only prep to %s", partition.ID, readOnlyDevice.MappedDevice)
+
+	return
+}
+
+// CleanupVerityDevice removes the device mapper linear mapping, but leaves the backing device unchanged
+func (v *VerityDevice) CleanupVerityDevice() {
+	stdout, stderr, err := shell.Execute("dmsetup", "remove", v.MappedName)
+	if err != nil {
+		err = fmt.Errorf("Unable to clean up device mapper device '%s' '%s': %w", stdout, stderr, err)
+		return
```

Comment:
You did not declare the error as a return value, hence you're not returning it here.

Reply:
Good catch, I'm going to both print this and return it since this would generally be a defered function

---
In reply to: [552719407](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552719407) [](ancestors = 552719407)

Reply:
As an aside, we need to go through and handle failures in defer calls.
```
defer func() {
    if deferError := deferedFunction(); deferError != nil {
        err = deferError
    }
}
```
This will set the out functions err value. I've noticed we don't handle unmounts reliably sometimes and never report or handle it.

---
In reply to: [552952854](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552952854) [](ancestors = 552952854,552719407)